File: fuzzer.jl
using Random
using Printf

function create_seed_file(seed_file::String) #pass in the name of our seed file because so we can name it whatever
    default_seed = "Initial Seed Data"
    open(seed_file, "w") do io
        write(io, default_seed)
    end
    @warn "Created a default seed file with content: $default_seed"
end

function fuzz_this_thing(prng_seed::Int, iterations::Int, seed_file::String)
    if !isfile(seed_file)
        create_seed_file(seed_file)
        println("Created seed file.")
        printstyled(io, "seed file created: $seed_file", color=:red)
    end

    
    data = Vector{UInt8}(read(seed_file)) # read seed file into a mutable byte array

    rng = MersenneTwister(prng_seed) #julia docs say this is a quick prng
    #however, may as well try another one for speed commparison

    for i in 1:iterations
        for j in eachindex(data)
            if rand(rng) < 0.13  # 13% chance to mutate a byte
                data[j] = rand(rng, UInt8)  # random between 0x00 and 0xFF
            end
        end

        if i % 500 == 0 # Every 500 iterations, extend the data buffer with 10 random bytes
            append!(data, [rand(rng, UInt8) for _ in 1:10]) 
        end
    end
    # Write the fuzzed data to standard output
    write(stdout, data) # what is the difference here to println() ?
end

function main()
    if length(ARGS) != 2 # parse arguments
        @error "Usage: julia fuzzer.jl <prng_seed> <iterations>"
        exit(1)
    end

    try
        prng_seed = parse(Int, ARGS[1])
        iterations = parse(Int, ARGS[2])

        fuzz_this_thing(prng_seed, iterations, "_seed_") #call the fuzzer
    catch e
        @error "Both arguments <prng_seed> and <iterations> must be integers." exception=(e, catch_backtrace())
        exit(1)
        #i am running into this error if I delete the seed file and run just the julia file
        #however, if I run it again after the seed is created, it works. ??
    end
end

# Run the program if executed as a script
if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: fuzzer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
function createSeedFile() {
    var defaultSeed = Buffer.from("InitialSeedData");
    fs.writeFileSync("_seed_", defaultSeed);
    console.error("Created a default seed file with content: ".concat(defaultSeed.toString()));
}
function fuzzThisThing(prngSeed, iterations) {
    // Create seed file if it doesn't exist
    if (!fs.existsSync("_seed_")) {
        createSeedFile();
    }
    var data = Buffer.from(fs.readFileSync("_seed_")); // expliciting typing this is necessary for concat later
    //needs to be let, not const bc it will update
    // Set up PRNG
    var random = seededRandom(prngSeed);
    // Fuzzing logic
    for (var i = 0; i < iterations; i++) {
        for (var j = 0; j < data.length; j++) {
            if (random() < 0.13) { // 13% chance of mutation
                data[j] = Math.floor(random() * 256); // Random byte
            }
        }
        if ((i + 1) % 500 === 0) { // Every 500 iterations, extend the data
            var additionalData = Buffer.alloc(10); // Create a buffer of size 10
            for (var k = 0; k < 10; k++) {
                additionalData[k] = Math.floor(random() * 256);
            }
            // concat original data with new data. TS has no dynamic resizing, need to create new arrays
            data = Buffer.concat([data, additionalData]);
        }
    }
    // Output the fuzzed data
    process.stdout.write(data);
    console.log(data.length);
}
function seededRandom(seed) {
    // Simple seeded random number generator
    return function () {
        var x = Math.sin(seed++) * 10000;
        return x - Math.floor(x);
    };
}
function main() {
    var args = process.argv.slice(2);
    if (args.length !== 2) {
        console.error("Usage: node fuzzer.js <prng_seed> <iterations>");
        process.exit(1);
    }
    var prngSeed = parseInt(args[0], 10);
    var iterations = parseInt(args[1], 10);
    if (isNaN(prngSeed) || isNaN(iterations)) {
        console.error("Both <prng_seed> and <iterations> must be valid integers.");
        process.exit(1);
    }
    console.log(iterations);
    fuzzThisThing(prngSeed, iterations);
}
main();
